### ISSUE TITLE
Encoding 'utf-8' errors

### ISSUE THREAD

**Issue by bschwab003** [oracle/python-cx_Oracle#122]
We are running into the following error.  

    'utf-8' codec can't decode byte 0x90 in position 9: invalid start byte

While its pretty obvious from the error message, we are setting/using UTF8 as our encoding type in our connection.

    dsn = cx_Oracle.makedsn(host=host, port=port, service_name=service_name)
    return cx_Oracle.connect(user, password, dsn, encoding="UTF8")

We are looking into whether or not we can identify the record and/or records containing the character(s) we cannot decode......but until we can specifically identify the characeter(s), do you have any guidance/thoughts on the following: 

1. Should we be using something different than "UTF8"?  Based on my research, in this repository and elsewhere, my gut tells me we should be using AL32UTF8 as it is more encompassing than UTF8 because UTF8 is only compatible up to Unicode Version 3.1 and AL32UTF8 would include the updates made in 3.2+?   

2.  Is there a way of "ignoring" encoding errors that doesn't involve using fetchone() to know exactly which record errored out and trying to handle it explicitly?  We typically pull data using fetchmany() with a size of 50K.  

Thanks,

Thanks,
Brian

**Comment by anthony-tuininga** [comment]
1. You should use "UTF-8", which is the IANA name and is equivalent to AL32UTF8.

2. Not currently. If you can generate a test case I can take a closer look at how to track this sort of thing down.

**Comment by bschwab003** [comment]
Thank you.

1.  So what encoding is cx_Oracle using when I enter "UTF8" instead of "UTF-8" if it's not AL32UTF8?

2.  I will certainly provide if/when we identify the specific record/field/characters.

**Comment by anthony-tuininga** [comment]
You're welcome.

1. cx_Oracle uses the Oracle character set name if it cannot find an encoding using the IANA name. So in this case, the Oracle character set UTF8 would be used (the less capable encoding as you guessed).

2. Thanks.

**Comment by bschwab003** [comment]
Re-opening this to ask a follow up question.  We received the following error again during extraction...

`'utf-8' codec can't decode byte 0xf7 in position 1: invalid start byte`

I thought we had this resolved per the thread above, but it has surfaced again.  

We resolved by inserting the following line at the beginning of the program..

`os.environ["NLS_LANG"] = "AMERICAN_AMERICA.AL32UTF8"`

After receiving this error, we also tried re-running it by setting the encoding as follows but we still received the same error.

`cx_Oracle.connect(user, password, dsn, encoding="UTF-8", nencoding="UTF-8")`

I am trying to see if I can track down the exact row that is causing the issue....but we are reading in data in 50K chunks so I am trying to narrow it down....

But could this really just be a "bad" character that UTF-8 can't decode?    And if so, is there anyway to ignore the error but still get the chunk, maybe by replacing it with a # or something?
